checkout-core/src/main/java/com/adyen/checkout/core/util/FileDownloader.kt

```diff
@@ -0,0 +1,112 @@
+import android.os.Build
+import android.provider.MediaStore
+import androidx.annotation.RequiresApi
+import androidx.annotation.RequiresPermission
+import androidx.annotation.RestrictTo
+import com.adyen.checkout.core.log.LogUtil
+import com.adyen.checkout.core.log.Logger
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.io.File
+import java.io.FileOutputStream
+import java.net.MalformedURLException
+import java.net.URL
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+@Suppress("LongParameterList")
+class FileDownloader {
+
+    @RequiresPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+    suspend fun download(
+        context: Context,
+        stringUrl: String,
+        fileName: String,
+        filePath: String,
+        mimeType: String? = null,
+        notifyState: () -> Unit
+    ) {
+        val url = stringUrl.toULR() ?: return
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+            downloadAndSaveApi29AndAbove(context, url, fileName, filePath, mimeType ?: "*/*", notifyState)
+        } else {
+            downloadAndSaveApi28AndBelow(context, url, fileName, filePath, notifyState)
+        }
+    }
+
+    @RequiresPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+    private suspend fun downloadAndSaveApi28AndBelow(
+        context: Context,
+        url: URL,
+        fileName: String,
+        filePath: String,
+        notifyState: () -> Unit
+    ) = withContext(Dispatchers.IO) {
+        val imageFile = File(context.getExternalFilesDir(filePath), fileName)
+        url.openStream().use { input ->
+            FileOutputStream(imageFile).use { output ->
+                input.copyTo(output)
+                withContext(Dispatchers.Main) { notifyState() }
+            }
+        }
+        MediaStore.Images.Media.insertImage(
+            context.contentResolver,
+            imageFile.absolutePath,
+            fileName,
+            null
+        )
+    }
+
+    @RequiresApi(Build.VERSION_CODES.Q)
+    private suspend fun downloadAndSaveApi29AndAbove(
+        context: Context,
+        url: URL,
+        fileName: String,
+        filePath: String,
+        mimeType: String,
+        notifyState: () -> Unit
+    ) = withContext(Dispatchers.IO) {
+        val contentValues = ContentValues().apply {
+            put(MediaStore.MediaColumns.DISPLAY_NAME, fileName)
+            put(MediaStore.MediaColumns.MIME_TYPE, mimeType)
+            put(MediaStore.MediaColumns.RELATIVE_PATH, filePath)
+        }
+        val resolver = context.contentResolver
+        val uri = resolver.insert(MediaStore.Downloads.EXTERNAL_CONTENT_URI, contentValues)
+        if (uri != null) {
+            url.openStream().use { input ->
+                resolver.openOutputStream(uri).use { output ->
+                    output?.let {
+                        input.copyTo(it, DEFAULT_BUFFER_SIZE)
+                        withContext(Dispatchers.Main) { notifyState() }
```

Comment:
Also don't forget to always return a value, either an exception or a successful result. For example, `if (uri == null)` and `if(output == null)` are both completely ignored